src/plugins/intel_npu/src/backend/src/zero_infer_request.cpp

```diff
@@ -40,18 +40,69 @@ void check_level_zero_attributes_match(const IONodeDescriptor& nodeDescriptor,
             "Given: " + std::to_string(ovDimensions.size()));
     }
 
-    for (size_t index = 0; index < ovDimensions.size(); ++index) {
-        if (ovDimensions[index] != zeDescriptor.info.dims[index] && !nodeDescriptor.originalShape.is_dynamic()) {
-            OPENVINO_THROW("Shape mismatch for parameter " + name);
-        }
-    }
     for (size_t index = ovDimensions.size(); index < ZE_MAX_GRAPH_ARGUMENT_DIMENSIONS_SIZE; ++index) {
         if (zeDescriptor.info.dims[index] != 0 && zeDescriptor.info.dims[index] != 1) {
             OPENVINO_THROW("Shape mismatch for parameter " + name);
         }
     }
 }
 
+size_t get_batch_size_for_node(const IONodeDescriptor& nodeDescriptor,
+                               const ZeroExecutor::ArgumentDescriptor& zeDescriptor) {
+    const std::vector<size_t>& ovDimensions = nodeDescriptor.originalShape.get_shape();
+    switch (zeDescriptor.info.deviceLayout) {
+    case ZE_GRAPH_ARGUMENT_LAYOUT_NCHW:
+    case ZE_GRAPH_ARGUMENT_LAYOUT_NHWC:
+    case ZE_GRAPH_ARGUMENT_LAYOUT_NCDHW:
+    case ZE_GRAPH_ARGUMENT_LAYOUT_NDHWC:
+    case ZE_GRAPH_ARGUMENT_LAYOUT_NC:
+        if ((ovDimensions[0] == zeDescriptor.info.dims[0]) && (ovDimensions[0] != 1)) {
+            OPENVINO_THROW("Batching on the plugin is not used, batching is handled by the compiler");
+        } else {
+            return ovDimensions[0];
+        }
+        break;
+    default:
+        OPENVINO_THROW("Batching on the plugin is working only when batching is found on 0th dimension");
+    }
+
+    return 1;
+}
+
+std::optional<size_t> get_batch_size(
+    const NetworkMetadata& metadata,
+    const std::unordered_map<std::string, ZeroExecutor::ArgumentDescriptor>& executorInputDescriptors,
+    const std::unordered_map<std::string, ZeroExecutor::ArgumentDescriptor>& executorOutputDescriptors) {
+    std::set<size_t> batch_size;
+
+    for (const std::string& inputName : metadata.inputNames) {
+        if (!executorInputDescriptors.count(inputName)) {
```

Comment:
I would appreciate if we could avoid doing this check twice. Also, do you believe an assertion is more adequate here?

Reply:
Just removed this check from here for now.

Reply:
Hmm..I believe removing the check might be a slightly worse approach. Assume
`config.get<BATCH_MODE>() != ov::intel_npu::BatchMode::COMPILER`
then `get_batch_size` will be called before the checks from the `ZeroInferRequest` ctor took place (i.e. `!executorInputDescriptors.count(inputName)`).

If there is a mismatch in the input identifier used, then `executorInputDescriptors.at(inputName)` from `get_batch_size` will fail with a lame error, such as:
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  map::at
Program terminated with signal: SIGSEGV
```

Instead of the more suggestive message:
`Invalid graph output descriptor key: SomeName`

So a tiny bit of debugability is lost as a consequence. What I was looking for was to move the `!executorInputDescriptors.count(inputName)` check before the `get_batch_size` call inside the `ZeroInferRequest` ctor in order to avoid duplicates and throw a suggestive message for both branches.

This is not of great relevance however, feel free to overlook this as well.